core/keyserver/epochs.go

```diff
@@ -34,27 +36,55 @@ var (
 	maxPageSize = int32(2048) // 8MB
 )
 
-// GetEpoch returns a list of mutations paged by epoch number.
+// GetLatestEpoch returns the latest epoch. The current epoch tracks the SignedLogRoot.
+func (s *Server) GetLatestEpoch(ctx context.Context, in *pb.GetLatestEpochRequest) (*pb.Epoch, error) {
+	// Lookup log and map info.
+	d, err := s.domains.Read(ctx, in.DomainId, false)
+	if err != nil {
+		glog.Errorf("adminstorage.Read(%v): %v", in.DomainId, err)
+		return nil, status.Errorf(codes.Internal, "Cannot fetch domain info")
+	}
+
+	// Fetch latest revision.
+	sth, err := s.latestLogRoot(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+	currentEpoch, err := latestRevision(sth)
+	if err != nil {
+		glog.Errorf("latestRevision(log %v, sth%v): %v", d.LogID, sth, err)
+		return nil, err
+	}
+
+	return s.getEpochByRevision(ctx, d, in.GetFirstTreeSize(), currentEpoch)
+}
+
+// GetEpoch returns a the requested epoch.
 func (s *Server) GetEpoch(ctx context.Context, in *pb.GetEpochRequest) (*pb.Epoch, error) {
 	if err := validateGetEpochRequest(in); err != nil {
 		glog.Errorf("validateGetEpochRequest(%v): %v", in, err)
 		return nil, status.Error(codes.InvalidArgument, "Invalid request")
 	}
 
 	// Lookup log and map info.
-	domain, err := s.domains.Read(ctx, in.DomainId, false)
+	d, err := s.domains.Read(ctx, in.DomainId, false)
 	if err != nil {
 		glog.Errorf("adminstorage.Read(%v): %v", in.DomainId, err)
 		return nil, status.Errorf(codes.Internal, "Cannot fetch domain info")
 	}
 
+	return s.getEpochByRevision(ctx, d, in.GetFirstTreeSize(), in.GetEpoch())
+
+}
+
+func (s *Server) getEpochByRevision(ctx context.Context, d *domain.Domain, firstTreeSize, revision int64) (*pb.Epoch, error) {
 	// Get signed map root by revision.
 	resp, err := s.tmap.GetSignedMapRootByRevision(ctx, &tpb.GetSignedMapRootByRevisionRequest{
-		MapId:    domain.MapID,
-		Revision: in.Epoch,
+		MapId:    d.MapID,
+		Revision: revision,
 	})
 	if err != nil {
-		glog.Warningf("GetSignedMapRootByRevision(%v, %v): %v", domain.MapID, in.Epoch, err)
+		glog.Warningf("GetSignedMapRootByRevision(%v, %v): %v", d.MapID, revision, err)
```

Comment:
This style of error message might make it hard to track down where they occurred. Might want to prefix the text with the rpc method name.

Reply:
Good idea.